### PR TITLE
Reverting ceph-disk removal

### PR DESCRIPTION
**Pull request by delfinogomes**
Adding ceph-disk to continue to use the same version of puppet-ceph

### CODE REVIEW

**Comment by lsvinicius-wrs** [comment]
Why are there other files/folders being commited besides ceph-disk ? For example ceph-detect-init and udev.

**Comment by delfinogomes** [comment]
ceph-disk uses ceph-detect-init on https://github.com/delfinogomes/stx-ceph/blob/54277dfb9d84e348440e767d62dc6ede78df0edd/src/ceph-disk/ceph_disk/main.py#L1044

Udev is needed because ceph-disk relies on these rules as documented on the documentation:

> The ceph-volume tool aims to be a single purpose command line tool to deploy logical volumes as OSDs, trying to maintain a similar API to ceph-disk when preparing, activating, and creating OSDs.
It deviates from ceph-disk by not interacting or relying on the udev rules that come installed for Ceph. These rules allow automatic detection of previously setup devices that are in turn fed into ceph-disk to activate them.

**Review by rchurch-wrs — COMMENTED** [review]
Seems reasonable to me. I'm assuming that this a one or more commits cherry-picked from  a previous branch and squashed? Can you specify the branch and  list the commits in the commit message. Use 'git log --abbrev-commit --pretty=oneline' for the formatting of the commit list. It's fairly compact.

File: src/ceph-disk/ceph_disk/main.py [hunk 1]
```diff
@@ -0,0 +1,5828 @@
...
+                'ceph-osd',
+                '--no-mon-config',
+                '--cluster', cluster,
+                '--mkfs',
+                '-i', osd_id,
+                '--monmap', monmap,
+                '--osd-data', path,
+                '--osd-journal', os.path.join(path, 'journal'),
+                '--osd-uuid', fsid,
+                '--osd-objectstore', 'filestore',
```
Comment (lsvinicius-wrs) [inline src/ceph-disk/ceph_disk/main.py:3218]: I don't see this parameter in https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/ceph-disk/ceph_disk/main.py#L3215

why is this needed ?

File: src/ceph-disk/ceph_disk/main.py [hunk 2]
```diff
@@ -0,0 +1,5828 @@
...
+    else:
+        fscheck, _, _ = command(
+            [
+                'blkid',
+                '-s',
+                'TYPE',
+                dev,
+            ],
+        )
+        if 'TYPE' in fscheck:
```
Comment (lsvinicius-wrs) [inline src/ceph-disk/ceph_disk/main.py:4350]: This if is different from https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/ceph-disk/ceph_disk/main.py#L4347

Why ?

File: src/ceph-disk/ceph_disk/main.py [hunk 3]
```diff
@@ -0,0 +1,5828 @@
...
+
+
+class PrepareJournal(PrepareSpace):
+
+    def __init__(self, args):
+        self.name = 'journal'
+        # ceph-osd now tries to identify the OSD type by checking 
+        # the disk, but at this stage we did not create anything
+        # because of that we assume the values for a Filestore OSD
+        self.allows_journal = True
```
Comment (lsvinicius-wrs) [inline src/ceph-disk/ceph_disk/main.py:2478]: In https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/ceph-disk/ceph_disk/main.py#L2478 it is using check_journal_reqs to fill these values.

A couple questions here:

The issue is there are no created OSDs when check_journal_reqs runs, right ?

What is the operation being made that raises the issue ? Is is trying to create an OSD ?

File: systemd/CMakeLists.txt [hunk 4]
```diff
@@ -31,5 +31,6 @@ install(FILES
   ceph-mds.target
   ceph-radosgw.target
   ceph-rbd-mirror.target
+  ceph-disk@.service
```
Comment (lsvinicius-wrs) [inline systemd/CMakeLists.txt:34]: What is the reason to create this service ?

File: systemd/ceph-disk@.service [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[Unit]
```
Comment (lsvinicius-wrs) [inline systemd/ceph-disk@.service:1]: What is the reason to create this service ?

File: systemd/ceph-disk@.service [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+Description=Ceph disk activation: %f
+After=local-fs.target
+Wants=local-fs.target
+
+[Service]
+Type=oneshot
+KillMode=none
+Environment=CEPH_DISK_TIMEOUT=10000
+ExecStart=/bin/sh -c 'timeout $CEPH_DISK_TIMEOUT flock /var/lock/ceph-disk-$(basename %f) /usr/sbin/ceph-disk --verbose --log-stdout trigger --sync %f'
```
Comment (lsvinicius-wrs) [inline systemd/ceph-disk@.service:10]: I believe the logs displayed by "--log-stdout" won't be saved, right ? It would be interesting if we could log the messages to a single file

File: src/pybind/mgr/status/module.py [hunk 7]
```diff
@@ -183,7 +183,7 @@ def handle_fs_status(self, cmd):
         output += "\n" + standby_table.get_string() + "\n"
 
         if len(mds_versions) == 1:
-            output += "MDS version: {0}".format(list(mds_versions))[0])
```
Comment (lsvinicius-wrs) [inline src/pybind/mgr/status/module.py:186]: In 13.2.2 it is using *mds_versions.keys()* https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/pybind/mgr/status/module.py#L243

I see in 14.2.22 https://github.com/starlingx-staging/stx-ceph/blob/83c70bf033cd6f7b9ea1a1f96023dcfcb912f977/src/pybind/mgr/status/module.py#L186 this has changed but syntax is wrong.

How did you verify this syntax was wrong ?

File: systemd/ceph-disk@.service (hunk 6 above)
Comment (lsvinicius-wrs) [inline systemd/ceph-disk@.service:10]: I believe the logs displayed by "--log-stdout" won't be saved, right ? It would be interesting if we could log the messages to a single file

File: src/ceph-disk/ceph_disk/main.py (hunk 3 above)
Comment (lsvinicius-wrs) [inline src/ceph-disk/ceph_disk/main.py:2479]: In https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/ceph-disk/ceph_disk/main.py#L2478 it is using check_journal_reqs to fill these values.

A couple questions here:

The issue is there are no created OSDs when check_journal_reqs runs, right ?

What is the operation being made that raises the issue ? Is is trying to create an OSD ?

File: src/ceph-disk/ceph_disk/main.py (hunk 1 above)
Comment (lsvinicius-wrs) [inline src/ceph-disk/ceph_disk/main.py:3219]: I don't see this parameter in https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/ceph-disk/ceph_disk/main.py#L3215 why is this needed ?

File: src/ceph-disk/ceph_disk/main.py (hunk 2 above)
Comment (lsvinicius-wrs) [inline src/ceph-disk/ceph_disk/main.py:4351]: This whole **if block** is different from https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/ceph-disk/ceph_disk/main.py#L4347 Why ?

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -464,7 +464,9 @@ if(NOT CEPH_BUILD_VIRTUALENV)
 endif()
```
Comment (lsvinicius-wrs) [inline src/CMakeLists.txt:464]: These changes are needed to compile ceph-disk, right ?

Comment (lsvinicius-wrs) [reply src/CMakeLists.txt:464]: Just checked this is the same as in https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/CMakeLists.txt#L773

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -615,6 +617,7 @@ install(PROGRAMS
 
 install(PROGRAMS
   ${CMAKE_SOURCE_DIR}/src/ceph-create-keys
+#  ${CMAKE_SOURCE_DIR}/src/ceph-disk
```
Comment (lsvinicius-wrs) [inline src/CMakeLists.txt:620]: This is a comment, it is not required to port this from https://github.com/starlingx-staging/stx-ceph/blob/ddfb3f76cc29f499fa13730e8ff6335d80cdc3b2/src/CMakeLists.txt#L964 but I think it is okay to leave it as it is.

**Review by lsvinicius-wrs — COMMENTED** [review]
There are some changes I'd like to understand better.

File: src/pybind/mgr/status/module.py (hunk 7 above)
Comment (delfinogomes) [reply src/pybind/mgr/status/module.py:186]: You can check this with 'ceph -s' command.

File: src/ceph-disk/ceph_disk/main.py (hunk 1 above)
Comment (delfinogomes) [reply src/ceph-disk/ceph_disk/main.py:3219]: The default objectstore is now bluestore. So I made explicit that I want to use filestore.

File: src/ceph-disk/ceph_disk/main.py (hunk 2 above)
Comment (delfinogomes) [reply src/ceph-disk/ceph_disk/main.py:4351]: this is missing this commit https://github.com/starlingx-staging/stx-ceph/commit/9cb6e15c71a9b494abda1d59b135b256933ea00e#diff-27b2d250f9ffa75e2f420b4fc9fde93a282d99fb7f89cd6a778df4e67b8975dd

File: systemd/ceph-disk@.service (hunk 5 above)
Comment (delfinogomes) [reply systemd/ceph-disk@.service:1]: While I was reverting ceph-disk, this was one of the files referenced in ceph.spec.in the I got it back as well.

File: systemd/ceph-disk@.service (hunk 6 above)
Comment (delfinogomes) [reply systemd/ceph-disk@.service:10]: I think we can improve that in a separate task, if needed.

File: src/ceph-disk/ceph_disk/main.py (hunk 1 above)
Comment (lsvinicius-wrs) [reply src/ceph-disk/ceph_disk/main.py:3219]: This is weird because there is a condition to enter this block checking if osd type is filestore.
I don't think this parameter should be added. Were there issues that were solved by adding this parameter ?

File: src/ceph-disk/ceph_disk/main.py (hunk 3 above)
Comment (delfinogomes) [reply src/ceph-disk/ceph_disk/main.py:2479]: To summarize ceph-disk makes three calls to ceph-osd and each of these calls should be delegated to the functions wants_journal, allows_journal and needs_journal of the ObjectStore class.

I had some problems making these calls to work, probably because just ceph-disk used this before its removal.

What I did to solve this was checking the implementations of these functions on FileStore.h
https://github.com/starlingx-staging/stx-ceph/blob/83c70bf033cd6f7b9ea1a1f96023dcfcb912f977/src/os/filestore/FileStore.h#L488-L496

And used those values directly int the class PrepareJournal that is only used by PrepareFilestore on ceph-disk.

**Comment by delfinogomes** [comment]
@rchurch-wrs I did not cherry-pick these changes.
My method was to check the history for the commits that removed ceph-disk, checked the deleted/modified files and then I got these files from branch stx/13.2.2

File: src/ceph-disk/ceph_disk/main.py (hunk 1 above)
Comment (delfinogomes) [reply src/ceph-disk/ceph_disk/main.py:3219]: I don't mean on ceph-disk, I mean the default object store on the command being called, which is ceph-osd.

File: src/ceph-disk/ceph_disk/main.py (hunk 2 above)
Comment (delfinogomes) [reply src/ceph-disk/ceph_disk/main.py:4351]: Cherry picked the missing commit, tks.

**Comment by lsvinicius-wrs** [comment]
I think this is fine to be merged now.

**Review by lsvinicius-wrs — COMMENTED** [review]
I think this is fine to be merged now.
